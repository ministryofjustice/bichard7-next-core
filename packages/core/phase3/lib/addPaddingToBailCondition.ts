const lineLength = 50

const addPaddingToBailCondition = (bailCondition: string): string => {
  const patternToMatchLine = new RegExp(`\\s*(.{1,${lineLength}})\\s+`, "gm")
  const lines = Array.from(bailCondition.concat(" ").matchAll(patternToMatchLine), (match) => match[1].trim())

  return lines
    .map((line, index) => {
      const padding = index < lines.length - 1 ? " ".repeat(lineLength - line.length) : ""
      return line + padding
    })
    .join("")
}

export default addPaddingToBailCondition
