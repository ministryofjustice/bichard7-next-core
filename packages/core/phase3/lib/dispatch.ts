import { isError, type PromiseResult } from "@moj-bichard7/common/types/Result"
import type { Operation, PncUpdateDataset } from "../../types/PncUpdateDataset"
import { PncOperation } from "../../types/PncOperation"
import type PncUpdateRequestGenerator from "../types/PncUpdateRequestGenerator"
import normalDisposalGenerator from "./pncUpdateRequestGenerators/normalDisposalGenerator"
import disposalUpdateGenerator from "./pncUpdateRequestGenerators/disposalUpdateGenerator"
import penaltyNoticeChargeGenerator from "./pncUpdateRequestGenerators/penaltyNoticeChargeGenerator"
import remandGenerator from "./pncUpdateRequestGenerators/remandGenerator"
import sentenceDeferredGenerator from "./pncUpdateRequestGenerators/sentenceDeferredGenerator"
import type PncGatewayInterface from "../../types/PncGatewayInterface"
import type { PncApiError } from "../../lib/PncGateway"
import generatePncUpdateExceptionFromMessage from "../exceptions/generatePncUpdateExceptionFromMessage"

const pncUpdateRequestGenerator: { [T in PncOperation]: PncUpdateRequestGenerator<T> } = {
  [PncOperation.DISPOSAL_UPDATED]: disposalUpdateGenerator,
  [PncOperation.NORMAL_DISPOSAL]: normalDisposalGenerator,
  [PncOperation.PENALTY_HEARING]: penaltyNoticeChargeGenerator,
  [PncOperation.REMAND]: remandGenerator,
  [PncOperation.SENTENCE_DEFERRED]: sentenceDeferredGenerator
}

const dispatch = async <T extends PncOperation>(
  pncUpdateDataset: PncUpdateDataset,
  operation: Operation<T>,
  pncGateway: PncGatewayInterface
): PromiseResult<void> => {
  const pncUpdateRequest = pncUpdateRequestGenerator[operation.code as T](pncUpdateDataset, operation)
  if (isError(pncUpdateRequest)) {
    return pncUpdateRequest
  }

  const correlationId = pncUpdateDataset.AnnotatedHearingOutcome.HearingOutcome.Hearing.SourceReference.UniqueID
  const pncUpdateResult = await pncGateway.update(pncUpdateRequest, correlationId).catch((error: PncApiError) => error)

  if (isError(pncUpdateResult)) {
    for (const message of pncUpdateResult.messages) {
      pncUpdateDataset.Exceptions.push(generatePncUpdateExceptionFromMessage(message))
    }

    return pncUpdateResult
  }
}

export default dispatch
