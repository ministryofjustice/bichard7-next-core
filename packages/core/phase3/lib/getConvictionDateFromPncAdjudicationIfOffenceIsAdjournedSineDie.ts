import findPncCourtCase from "../../phase2/lib/findPncCourtCase"
import type { AnnotatedHearingOutcome, Offence } from "../../types/AnnotatedHearingOutcome"

const ADJOURNED_SINE_DIE_DISPOSAL_CODE = 2007

const getConvictionDateFromPncAdjudicationIfOffenceIsAdjournedSineDie = (
  aho: AnnotatedHearingOutcome,
  offence: Offence
) => {
  if (
    !offence.CriminalProsecutionReference.OffenceReasonSequence ||
    !aho.PncQuery ||
    aho.AnnotatedHearingOutcome.HearingOutcome.Case.PenaltyNoticeCaseReferenceNumber
  ) {
    return undefined
  }

  const courtCase = findPncCourtCase(aho, offence)
  const matchingOffence = courtCase?.offences.find(
    (pncOffence) =>
      pncOffence.offence.sequenceNumber === Number(offence.CriminalProsecutionReference.OffenceReasonSequence)
  )

  const areAllDisposals2007 =
    matchingOffence?.disposals &&
    matchingOffence.disposals.length > 0 &&
    matchingOffence.disposals.every((disposal) => disposal.type === ADJOURNED_SINE_DIE_DISPOSAL_CODE)

  return areAllDisposals2007 ? matchingOffence.adjudication?.sentenceDate : undefined
}

export default getConvictionDateFromPncAdjudicationIfOffenceIsAdjournedSineDie
