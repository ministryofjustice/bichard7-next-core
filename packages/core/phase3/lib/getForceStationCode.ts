import type { AnnotatedHearingOutcome } from "../../types/AnnotatedHearingOutcome"

const bichardThirdLevelCode = "YZ"

const getForceStationCode = (aho: AnnotatedHearingOutcome, useSpecialStationCode: boolean): string => {
  const forceOwner = aho.AnnotatedHearingOutcome.HearingOutcome.Case.ForceOwner
  const secondLevelCode = forceOwner?.SecondLevelCode
  const thirdLevelCode = useSpecialStationCode ? bichardThirdLevelCode : forceOwner?.ThirdLevelCode
  return secondLevelCode ? `${secondLevelCode}${thirdLevelCode}` : "0000"
}

export default getForceStationCode
