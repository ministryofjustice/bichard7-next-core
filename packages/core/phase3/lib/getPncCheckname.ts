import type { AnnotatedHearingOutcome } from "../../types/AnnotatedHearingOutcome"

const getPncCheckname = (aho: AnnotatedHearingOutcome) =>
  aho.AnnotatedHearingOutcome.HearingOutcome.Case.HearingDefendant.PNCCheckname?.split("/")[0].substring(0, 12) ?? null

export default getPncCheckname
