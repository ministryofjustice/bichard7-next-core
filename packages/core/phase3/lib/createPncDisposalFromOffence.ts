import type { AnnotatedHearingOutcome, Offence } from "../../types/AnnotatedHearingOutcome"
import type { PncDisposal } from "../../types/PncQueryResult"

import { createPncDisposalsFromResult } from "../../phase2/lib/createPncDisposalsFromResult"
import createPncDisposal from "../../phase2/lib/createPncDisposalsFromResult/createPncDisposal"
import findPncCourtCase from "../../phase2/lib/findPncCourtCase"
import isRecordableResult from "../../phase2/lib/isRecordableResult"

const ADJOURNED_SINE_DIE_DISPOSAL_CODE = 2007
const getConvictionDateFromPncAdjudicationIfOffenceIsAdjournedSineDie = (
  aho: AnnotatedHearingOutcome,
  offence: Offence
) => {
  if (
    !offence.CriminalProsecutionReference.OffenceReasonSequence ||
    !aho.PncQuery ||
    aho.AnnotatedHearingOutcome.HearingOutcome.Case.PenaltyNoticeCaseReferenceNumber
  ) {
    return undefined
  }

  const courtCase = findPncCourtCase(aho, offence)
  const matchingOffence = courtCase?.offences.find(
    (pncOffence) =>
      pncOffence.offence.sequenceNumber === Number(offence.CriminalProsecutionReference.OffenceReasonSequence)
  )

  const areAllDisposals2007 =
    matchingOffence?.disposals &&
    matchingOffence.disposals.length > 0 &&
    matchingOffence.disposals.every((disposal) => disposal.type === ADJOURNED_SINE_DIE_DISPOSAL_CODE)

  return areAllDisposals2007 ? matchingOffence.adjudication?.sentenceDate : undefined
}

const createPncDisposalFromOffence = (aho: AnnotatedHearingOutcome, offence: Offence): PncDisposal[] => {
  const results = offence.Result
  let pncDisposals: PncDisposal[] = []

  let found3027 = false
  const adjournmentExists = results.some((result) => result.ResultClass?.includes("Adjournment"))
  let disposalFor2060Result: null | PncDisposal[] = null
  let found2050Result = false
  let found2063Result = false
  let converted2060Result = false

  results
    .filter(isRecordableResult)
    .sort((a, b) => a.CJSresultCode - b.CJSresultCode)
    .forEach((result) => {
      const disposalCode = result.PNCDisposalType
      const resultCode = result.CJSresultCode
      let ignore2063Disposal = false
      found3027 ||= disposalCode === 3027
      const generatedDisposals = createPncDisposalsFromResult(result)

      if (disposalCode === 2060 && disposalFor2060Result == null) {
        disposalFor2060Result = generatedDisposals
      } else if (disposalCode === 2050) {
        found2050Result = true
      } else if (disposalCode === 2063) {
        if (resultCode === 2060) {
          converted2060Result = true
          disposalFor2060Result = generatedDisposals
        } else {
          found2063Result = true
        }

        if (converted2060Result && found2063Result) {
          ignore2063Disposal = true
        }
      }

      if ((disposalCode !== 3052 || !adjournmentExists) && !ignore2063Disposal) {
        pncDisposals.push(...generatedDisposals)
      }
    })

  if (disposalFor2060Result && (found2050Result || found2063Result) && pncDisposals.length == 2) {
    pncDisposals = disposalFor2060Result
  }

  if (found3027 || adjournmentExists) {
    return pncDisposals
  }

  const convictionDate = getConvictionDateFromPncAdjudicationIfOffenceIsAdjournedSineDie(aho, offence)
  if (convictionDate) {
    pncDisposals.push(
      createPncDisposal(
        3027,
        undefined,
        undefined,
        undefined,
        undefined,
        convictionDate,
        undefined,
        undefined,
        undefined
      )
    )
  }

  return pncDisposals
}

export default createPncDisposalFromOffence
