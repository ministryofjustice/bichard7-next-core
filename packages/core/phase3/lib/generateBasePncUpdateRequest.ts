import type { PncUpdateDataset } from "@moj-bichard7/common/types/PncUpdateDataset"

import getForceStationCode from "./getForceStationCode"
import getPncCheckname from "./getPncCheckname"
import preProcessPncIdentifier from "./preProcessPncIdentifier"

const generateBasePncUpdateRequest = (pncUpdateDataset: PncUpdateDataset) => ({
  croNumber: pncUpdateDataset.AnnotatedHearingOutcome.HearingOutcome.Case.HearingDefendant.CRONumber ?? null,
  forceStationCode: getForceStationCode(pncUpdateDataset, true),
  pncCheckName: getPncCheckname(pncUpdateDataset),
  pncIdentifier: preProcessPncIdentifier(
    pncUpdateDataset.AnnotatedHearingOutcome.HearingOutcome.Case.HearingDefendant.PNCIdentifier
  )
})

export default generateBasePncUpdateRequest
