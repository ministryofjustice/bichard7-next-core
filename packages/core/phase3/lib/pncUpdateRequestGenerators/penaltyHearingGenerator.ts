import { isError } from "@moj-bichard7/common/types/Result"

import type PncUpdateRequestGenerator from "../../types/PncUpdateRequestGenerator"

import formatDateSpecifiedInResult from "../../../lib/createPncDisposalsFromResult/formatDateSpecifiedInResult"
import { PncOperation } from "../../../types/PncOperation"
import generateBasePncUpdateRequest from "../generateBasePncUpdateRequest"
import getPncCourtCode from "../getPncCourtCode"
import { generateHearingsAdjudicationsAndDisposals } from "../hearingDetails/generateHearingsAdjudicationsAndDisposals"

const PENALTY_HEARING_TYPE = "P"

const penaltyHearingGenerator: PncUpdateRequestGenerator<PncOperation.PENALTY_HEARING> = (
  pncUpdateDataset,
  operation
) => {
  const hearing = pncUpdateDataset.AnnotatedHearingOutcome.HearingOutcome.Hearing

  const penaltyNoticeCaseRef =
    operation.data?.courtCaseReference ??
    pncUpdateDataset.AnnotatedHearingOutcome.HearingOutcome.Case.PenaltyNoticeCaseReferenceNumber

  if (!penaltyNoticeCaseRef) {
    // We generate penalty hearing operation only if PenaltyNoticeCaseReferenceNumber has value in Phase 2.
    // If we get this error something is very wrong.
    return new Error("Penalty notice case ref is missing")
  }

  const courtCode = getPncCourtCode(hearing.CourtHearingLocation, hearing.CourtHouseCode)
  if (isError(courtCode)) {
    return courtCode
  }

  return {
    operation: PncOperation.PENALTY_HEARING,
    request: {
      ...generateBasePncUpdateRequest(pncUpdateDataset),
      courtCode,
      hearingDate: formatDateSpecifiedInResult(hearing.DateOfHearing, true),
      hearingDetails: generateHearingsAdjudicationsAndDisposals(pncUpdateDataset, penaltyNoticeCaseRef),
      hearingType: PENALTY_HEARING_TYPE,
      penaltyNoticeCaseRef: penaltyNoticeCaseRef
    }
  }
}

export default penaltyHearingGenerator
