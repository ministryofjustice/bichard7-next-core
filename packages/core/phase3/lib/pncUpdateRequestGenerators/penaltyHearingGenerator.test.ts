import { isError } from "@moj-bichard7/common/types/Result"
import fs from "fs"
import path from "path"

import type { Offence, OffenceReason, Result } from "../../../types/AnnotatedHearingOutcome"
import type { Operation, PncUpdateDataset } from "../../../types/PncUpdateDataset"

import { lookupOrganisationUnitByCode } from "../../../lib/dataLookup"
import parsePncUpdateDataSetXml from "../../../phase2/parse/parsePncUpdateDataSetXml/parsePncUpdateDataSetXml"
import { PncOperation } from "../../../types/PncOperation"
import ResultClass from "../../../types/ResultClass"
import penaltyHearingGenerator from "./penaltyHearingGenerator"

jest.mock("../../../lib/dataLookup/lookupOrganisationUnitByCode")
const mockedLookupOrganisationUnitByCode = lookupOrganisationUnitByCode as jest.Mock

const createPncUpdateDataset = () => {
  const offence = {
    OffenceCategory: "ZZ",
    CriminalProsecutionReference: {
      OffenceReason: {
        __type: "NationalOffenceReason",
        OffenceCode: {
          __type: "NonMatchingOffenceCode",
          ActOrSource: "RT",
          Year: "88",
          Reason: 191,
          FullCode: "RT88191"
        }
      } as unknown as OffenceReason
    },
    AddedByTheCourt: true,
    ActualOffenceStartDate: {
      StartDate: new Date("2025-01-01")
    },
    Result: [
      {
        NextHearingDate: "2024-12-11T10:11:12.000Z",
        NextResultSourceOrganisation: {
          TopLevelCode: "B",
          SecondLevelCode: "01",
          ThirdLevelCode: "00",
          BottomLevelCode: "00",
          OrganisationUnitCode: "B000000"
        },
        PNCDisposalType: 2059,
        PNCAdjudicationExists: true,
        ResultClass: ResultClass.ADJOURNMENT_PRE_JUDGEMENT,
        ResultQualifierVariable: [{ Code: "LE" }]
      }
    ] as Result[]
  } as Offence

  const pncUpdateDataset = {
    AnnotatedHearingOutcome: {
      HearingOutcome: {
        Hearing: {
          DateOfHearing: new Date("2024-12-05"),
          CourtHearingLocation: {
            TopLevelCode: "B",
            SecondLevelCode: "01",
            ThirdLevelCode: "00",
            BottomLevelCode: "00",
            OrganisationUnitCode: "B000000"
          },
          CourtHouseName: "Magistrates' Courts London Croydon",
          CourtType: "MCA"
        },
        Case: {
          CourtCaseReferenceNumber: "97/1626/008395Q",
          ForceOwner: {
            TopLevelCode: "A",
            SecondLevelCode: "02",
            ThirdLevelCode: "BJ",
            BottomLevelCode: "01",
            OrganisationUnitCode: "A02BJ01"
          },
          HearingDefendant: {
            ArrestSummonsNumber: "1101ZD0100000410780J",
            BailConditions: ["This is a dummy bail condition."],
            Offence: [offence],
            RemandStatus: "CB",
            CRONumber: "123"
          },
          PTIURN: "01ZD0303208",
          PenaltyNoticeCaseReferenceNumber: "97/1626/008395Q"
        }
      }
    },
    Exceptions: [],
    PncOperations: []
  } as unknown as PncUpdateDataset

  return pncUpdateDataset
}

describe("penaltyHearingGenerator", () => {
  beforeEach(() => {
    mockedLookupOrganisationUnitByCode.mockRestore()
    mockedLookupOrganisationUnitByCode.mockImplementation(
      jest.requireActual("../../../lib/dataLookup/lookupOrganisationUnitByCode").default
    )
  })

  it("generates the operation request", () => {
    const filePath = path.join(__dirname, "../../../phase2/tests/fixtures/PncUpdateDataSet-with-single-PENHRG.xml")
    const inputXml = fs.readFileSync(filePath).toString()
    const pncUpdateDataset = parsePncUpdateDataSetXml(inputXml)
    if (isError(pncUpdateDataset)) {
      throw pncUpdateDataset
    }

    const result = penaltyHearingGenerator(
      pncUpdateDataset,
      pncUpdateDataset.PncOperations[0] as Operation<PncOperation.PENALTY_HEARING>
    )

    expect(result).toMatchSnapshot()
  })

  it("should return error when court case reference in operation and penalty notice case reference number in hearing outcome are undefined", () => {
    const pncUpdateDataset = createPncUpdateDataset()
    pncUpdateDataset.PncOperations = [
      {
        code: PncOperation.PENALTY_HEARING,
        data: undefined
      } as Operation<PncOperation.PENALTY_HEARING>
    ]
    pncUpdateDataset.AnnotatedHearingOutcome.HearingOutcome.Case.PenaltyNoticeCaseReferenceNumber = undefined

    const result = penaltyHearingGenerator(
      pncUpdateDataset,
      pncUpdateDataset.PncOperations[0] as Operation<PncOperation.PENALTY_HEARING>
    )

    expect(isError(result)).toBe(true)
    expect((result as Error).message).toBe("Penalty notice case ref is missing")
  })

  it("should return error when it fails to get PSA court code from court hearing location", () => {
    const mockedLookupOrganisationUnitByCode = lookupOrganisationUnitByCode as jest.Mock

    mockedLookupOrganisationUnitByCode.mockReturnValue({
      bottomLevelCode: "00",
      bottomLevelName: "",
      secondLevelCode: "20",
      secondLevelName: "West Midlands",
      thirdLevelCode: "BN",
      thirdLevelName: "Birmingham Youth Court (Steelehouse Lane)",
      thirdLevelPsaCode: "I'm not a number",
      topLevelCode: "B",
      topLevelName: "Magistrates' Courts"
    })

    const pncOperation = {
      code: PncOperation.PENALTY_HEARING
    } as Operation<PncOperation.PENALTY_HEARING>
    const pncUpdateDataset = createPncUpdateDataset()
    pncUpdateDataset.AnnotatedHearingOutcome.HearingOutcome.Hearing.CourtHouseCode = 4001

    const result = penaltyHearingGenerator(pncUpdateDataset, pncOperation)

    expect(isError(result)).toBe(true)
    expect((result as Error).message).toBe("PSA code 'I'm not a number' is not a number")
  })
})
