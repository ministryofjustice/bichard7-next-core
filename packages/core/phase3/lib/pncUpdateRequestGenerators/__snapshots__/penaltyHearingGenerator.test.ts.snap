// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`penaltyHearingGenerator generates the operation request 1`] = `
{
  "operation": "PENHRG",
  "request": {
    "courtCode": "2575",
    "croNumber": null,
    "forceStationCode": "01YZ",
    "hearingDate": "19122023",
    "hearingDetails": [
      {
        "courtOffenceSequenceNumber": "001",
        "offenceReason": "FA06001",
        "type": "ORDINARY",
      },
      {
        "hearingDate": "19122023",
        "numberOffencesTakenIntoAccount": "0000",
        "pleaStatus": "GUILTY",
        "type": "ADJUDICATION",
        "verdict": "GUILTY",
      },
      {
        "disposalQualifiers": "",
        "disposalQuantity": "                      00",
        "disposalText": "FAILED TO APPEAR WARRANT ISSUED",
        "disposalType": "4004",
        "type": "DISPOSAL",
      },
      {
        "courtOffenceSequenceNumber": "002",
        "offenceReason": "FA06001",
        "type": "ORDINARY",
      },
      {
        "hearingDate": "19122023",
        "numberOffencesTakenIntoAccount": "0000",
        "pleaStatus": "GUILTY",
        "type": "ADJUDICATION",
        "verdict": "GUILTY",
      },
      {
        "disposalQualifiers": "",
        "disposalQuantity": "                      00",
        "disposalText": "FAILED TO APPEAR WARRANT ISSUED",
        "disposalType": "4004",
        "type": "DISPOSAL",
      },
    ],
    "hearingType": "P",
    "penaltyNoticeCaseRef": "07/1577/020154M",
    "pncCheckName": "COLE",
    "pncIdentifier": "24/1234567X",
  },
}
`;
